cpp/src/arrow/io/interfaces.cc

```diff
@@ -138,6 +137,42 @@ Result<Iterator<std::shared_ptr<Buffer>>> MakeInputStreamIterator(
   return Iterator<std::shared_ptr<Buffer>>(InputStreamBlockIterator(stream, block_size));
 }
 
+Result<AsyncGenerator<std::shared_ptr<Buffer>>> MakeRandomAccessFileGenerator(
+    std::shared_ptr<RandomAccessFile> file, int64_t block_size) {
+  struct State {
+    explicit State(std::shared_ptr<RandomAccessFile> file_, int64_t block_size_)
+        : file(std::move(file_)), block_size(block_size_), position(0) {}
+
+    Status init() {
+      RETURN_NOT_OK(file->Seek(0));
+      // if seek worked this will also work.
+      total_size = file->GetSize().ValueOrDie();
+      return Status::OK();
```

Comment:
```suggestion
      return file->GetSize();
```

You are probably right but I'm not sure avoiding this if check is worth making this method slightly more confusing given it shouldn't really be called all that often.